crates/services/gas_price_service/src/v1/da_source_adapter/block_committer_costs.rs

```diff
@@ -6,57 +8,321 @@ use crate::v1::da_source_adapter::{
+    ) {
+        // given
+        let mut da_block_costs = test_da_block_costs();
+        let mock_api = MockBlockCommitterApi::new(Some(da_block_costs.clone()));
+        let mut block_committer =
+            BlockCommitterDaBlockCosts::new(mock_api, Some(da_block_costs.clone()));
+
+        // when
+        let actual = block_committer.request_da_block_cost().await.unwrap();
+
+        // then
+        assert_ne!(da_block_costs.blocks_range, actual.l2_block_range);
+    }
+
+    #[tokio::test]
+    async fn request_da_block_cost__when_response_is_none__then_error() {
+        // given
+        let mock_api = MockBlockCommitterApi::new(None);
+        let mut block_committer = BlockCommitterDaBlockCosts::new(mock_api, None);
+
+        // when
+        let result = block_committer.request_da_block_cost().await;
+
+        // then
+        assert!(result.is_err());
+        assert!(block_committer.last_raw_da_block_costs.is_none());
+    }
+
+    struct UnderflowingMockBlockCommitterApi {
+        value: Option<RawDaBlockCosts>,
+    }
+
+    impl UnderflowingMockBlockCommitterApi {
+        fn new(value: Option<RawDaBlockCosts>) -> Self {
+            Self { value }
+        }
+    }
+
+    #[async_trait::async_trait]
+    impl BlockCommitterApi for UnderflowingMockBlockCommitterApi {
+        async fn get_latest_costs(&self) -> DaBlockCostsResult<Option<RawDaBlockCosts>> {
+            Ok(self.value.clone())
+        }
+        async fn get_costs_by_seqno(
+            &self,
+            seq_no: u64,
+        ) -> DaBlockCostsResult<Option<RawDaBlockCosts>> {
+            // arbitrary logic to generate a new value
+            let mut value = self.value.clone();
+            if let Some(value) = &mut value {
+                value.sequence_number = seq_no;
+                value.blocks_range = value.blocks_range.end..value.blocks_range.end + 10;
+                value.da_block_height = value.da_block_height + 1u64.into();
+                value.total_cost -= 1;
+                value.total_size_bytes -= 1;
+            }
+            Ok(value)
+        }
+        async fn get_cost_bundles_by_range(
+            &self,
+            _: core::ops::Range<u64>,
+        ) -> DaBlockCostsResult<Vec<Option<RawDaBlockCosts>>> {
+            Ok(vec![self.value.clone()])
+        }
+    }
+
+    #[tokio::test]
+    async fn request_da_block_cost__when_underflow__then_error() {
+        // given
+        let da_block_costs = test_da_block_costs();
+        let expected = (&da_block_costs).into();
+        let mock_api = UnderflowingMockBlockCommitterApi::new(Some(da_block_costs));
+        let mut block_committer = BlockCommitterDaBlockCosts::new(mock_api, None);
+
+        // when
+        let actual = block_committer.request_da_block_cost().await.unwrap();
+        let result = block_committer.request_da_block_cost().await;
+
+        // then
+        assert!(result.is_err());
```

Comment:
I was gonna say we should check the specific error variant here but `anyhow` :(

Reply:
in my own experience, I tend to prefer `thiserror::Error` to `anyhow` by a lot, especially in codebases where each module has its own error type.